### ISSUE TITLE
Request: Support for IDurableClientFactory in external processes

### ISSUE THREAD

**Issue by JackSkylark** [microsoft/durabletask-netherite#49]
Is it reasonable to assume that there will be support for using an IDurableClientFactory in external processes?

**Comment by sebastianburckhardt** [comment]
Yes, that is a very reasonable assumption.

**Comment by gorillapower** [comment]
Currently im using the AzureStorageProvider and IDurableClientFactory to communicate between different durable apps with my application, is this possible today with NetheriteStorage provider? Or do all functions need to be within the same Taskhub?

**Comment by sebastianburckhardt** [comment]
Yes, it is possible to configure durable clients to connect to different task hubs. As long as the task hub is running somewhere (anywhere), a client can communicate with it to start orchestrations and query the status of orchestrations. This communication is going through the event hubs (client requests to the task hub, and responses from the task hub, are transmitted as event hubs messages).

There are some operational differences compared to the Azure Storage provider:

- A Netherite-based client is responsive only if the function app is running somewhere (anywhere). This is different from the Azure Storage client which can talk to Azure Storage even if the function app is not currently running - because Azure Storage is "always running".

- There is a slight difference when configuring the connection name for durable clients using a C# attribute, as documented [here](https://microsoft.github.io/durabletask-netherite/#/connectionnames?id=connection-names-for-durableclients).

**Comment by sebastianburckhardt** [comment]
It is supposed to work, if it does not that is a bug we need to fix.

**Comment by gorillapower** [comment]
Question1: Understood

Question2: It seems like when DI injects `IDurableClientFactory` in a constructor, its always defaulting to `AzureStorageProvider` even if i have the provider set to Netherite in the host.json config.  

Additionally, if I try create my own `DurableClientFactory` (using the attribute method or manually creating a DurableClientFactory instance) there is a null exception, where the `TaskHubParameters` seem to be null on the transport object.

DurableTask.Netherite\OrchestrationService\NetheriteOrchestrationService.cs:366
<img width="743" alt="image" src="https://user-images.githubusercontent.com/8879065/210975476-6b72447d-06e6-4ac4-b423-e9b1e82dbd57.png">

It seems this method is never called, to create the `TaskHubParameters` when manually creating a DurableClientFactory instance or from the attribute method, resulting in the null exception.

DurableTask.Netherite\OrchestrationService\NetheriteOrchestrationService.cs:260
<img width="618" alt="image" src="https://user-images.githubusercontent.com/8879065/210975862-46ae51cd-fdbd-4cc3-8f83-02622ef9bcc8.png">



Im not sure if this only because because Im using the 'Memory' emulation mode, I have not tested with a real connection string yet and Im not sure where the pipeline the CreateAsync() is called. However, I am able to see a NetheriteOrchestrator correctly initialising, I think this happens on startup though. But I am still needing to create my own custom client, with different connection parameters.

___
Code i am using to create the DurableClientFactory

``` csharp
services.AddSingleton((Func<IServiceProvider, IDurableClientFactory>)((serviceProvider) =>
{
    IOptions<DurableClientOptions> durableClientOptions = serviceProvider.GetService<IOptions<DurableClientOptions>>();
    IOptions<DurableTaskOptions> durableTaskOptions = serviceProvider.GetService<IOptions<DurableTaskOptions>>();
    ILoggerFactory loggerFactory = serviceProvider.GetService<ILoggerFactory>();
    IDurabilityProviderFactory configuredFactory = GetConfiguredProviderFactory(serviceProvider, durableTaskOptions);

    DurableClientFactory durableClientFactory = new DurableClientFactory(durableClientOptions, durableTaskOptions, configuredFactory, loggerFactory);
    
    return durableClientFactory;

}));
        

private static IDurabilityProviderFactory GetConfiguredProviderFactory(IServiceProvider serviceProvider, IOptions<DurableTaskOptions> durableTaskOptions)
{
    IEnumerable<IDurabilityProviderFactory> providerFactories = serviceProvider.GetServices<IDurabilityProviderFactory>();

    string providerName = GetConfiguredStorageProviderName(durableTaskOptions);

    IDurabilityProviderFactory selectedFactory = providerFactories.First(f => string.Equals(f.Name, providerName, StringComparison.OrdinalIgnoreCase));
    return selectedFactory;
}

public static string GetConfiguredStorageProviderName(IOptions<DurableTaskOptions> durableTaskOptions)
{
    bool storageTypeIsConfigured = durableTaskOptions.Value.StorageProvider.TryGetValue("type", out object storageType);
    string storageName = "AzureStorage"; //default provider name
    if (storageTypeIsConfigured)
    {
        storageName = storageType.ToString();
    }

    return storageName;
}

```

**Comment by gorillapower** [comment]
Just a follow up on the exception, short answer is that is indeed because of the MemoryStorageProvider I am seeing the exception, after switching to real EventHub connection strings I got my code to execute successfully. 

It doesnt look like MemoryStorageProvider would work across different TaskHubs hosts and/or across multiple IDurableClient instances in the way I am using it. Im not sure if thats even something you would be wanting to support in the future with the MemoryStorageProvider, I did attempt to dig deeper and see if it was possible to support my scenario but ran out of time unfortunately.

However, the issue still remains where the IDurableClientFactory still returns the AzureStorageProvider, even when Netherite is specified as the storage provider in the localsettings.json. This means any call to IDurableClientFactory.CreateClient() will always return the default AzureStorageProvider. I assume the behaviour should be to return the client based on the users explicit configuration?

**Comment by sebastianburckhardt** [comment]
> It doesnt look like MemoryStorageProvider would work across different TaskHubs hosts and/or across multiple IDurableClient instances in the way I am using it

Yes, you are right, the MemoryStorageProvider can simulate only a single taskhub, with both client and worker in a single process (because there aren't actually any storage queues, tables, or blobs, it is all just in memory). If you want to run a configuration with multiple task hubs, or with clients that run in a different process than the workers, you will have to use a real Azure Storage and Event Hubs.

> However, the issue still remains where the IDurableClientFactory still returns the AzureStorageProvider, even when Netherite is specified as the storage provider in the localsettings.json.

I still have to look into this.

**Comment by sebastianburckhardt** [comment]
> This means any call to IDurableClientFactory.CreateClient() will always return the default AzureStorageProvider.  

I can repro this now on my own, so I understand better what exactly your are observing, and I now have a starting point for figuring out a workaround or fix.

> I assume the behaviour should be to return the client based on the users explicit configuration?

I would think so also. But the whole story around using DI for clients is not really documented as far as I can tell (if you have a link to docs somewhere let me know) so it is hard to say whether this is a bug or just something that was never really supported. The code you are using makes sense but it is also much more complicated than what I think it should have to be. Either way, I think we need to find a simpler way to enable this kind of scenario and then document and support it.

**Comment by gorillapower** [comment]
The only thing I could find was this issue  ['Add documentation for managing orchestrations/entities from external clients'](https://github.com/Azure/azure-functions-durable-extension/issues/1600) in azure-functions-durable-extension repo

**Comment by sebastianburckhardt** [comment]
After a bit more experimentation (and using your code as a starting point, thank you @gorillapower) I finally got it to work. I had to add bit more logic to configure the client options.  

I have created a sample in the folders `samples/StandaloneClient`. This is currently in a draft PR in this branch: https://github.com/microsoft/durabletask-netherite/tree/pr/external-client-samples.

This now works by injecting  new class `ExternalClientFactory` during startup which is then used with the Azure Function DI mechanism to create the client. I am not sure if this is exactly the setup you had in mind, maybe you want to use the client without using the functions host (e.g. from a console host). That would require a somewhat different configuration.

For now, maybe this sample code can unblock you. Going forward, I think it would make sense for us to define the ExternalClientFactory class inside the extension so we can maintain it as needed, once we are sure it meets all the requirements.

**Comment by gorillapower** [comment]
Great! This is similar implementation to what I ended up doing. I have a singleton factory that is then injected just like that, which conditionally sets the `DurableClientFactory` (AzureStorage, Netherite, etc) based on  appsettings configuration, awesome.

Its worth reiterating for anyone who comes across this, the default behaviour for `services.AddDurableClientFactory()` and thus any DI reference to 'IDurableClientFactory' will still return the the default `AzureStorage` provider, use the `ExternalClientFactory` if you leverage any non default storage providers, thanks @sebastianburckhardt for the time and attention.

Just a thought, it might also be useful now to think about the differences/similarities of IDurableClientFactory and ExternalClientFactory?  Do you envisage that one day IDurableClientFactory would support non-standard storage providers or would ExternalClientFactory sit on its own? We should also keep in mind, IDurableClientFactory isnt documented yet...

**Comment by lilyjma** [comment]
Hi @JackSkylark / @gorillapower, thanks for your interest in this new feature! We're looking for ways to improve it further. Would it be possible to schedule a quick chat with you to discuss your experience using it? If yes, please share your email via a [LinkedIn message](https://www.linkedin.com/in/lilyjma0225/). (If you don't have LinkedIn, we can figure out another way.) Thanks again!

**Comment by gorillapower** [comment]
Hi @lilyjma I am unable to message on facebook, is there another way?
